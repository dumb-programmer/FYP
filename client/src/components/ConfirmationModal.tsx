import { Action, Content, Header, Modal } from "./modal";

export default function ConfirmationModal({ dialogRef, title, description, primaryBtnClass = "btn-error", primaryBtnText = "Delete", onPrimary, onCancel }: { dialogRef: React.MutableRefObject<HTMLDialogElement | null>, title: string, description: string, primaryBtnClass?: string, primaryBtnText?: string, onPrimary: () => void, onCancel: () => void }) {
    return <Modal dialogRef={dialogRef}>
        <Header>
            <h1 className="text-xl font-bold">
                {title}
            </h1>
        </Header>
        <Content>
            <p className="mt-4">
                {description}
            </p>
        </Content>
        <Action>
            <button data-testid="delete-confirmation-dialog-cancel-btn" className="btn" onClick={onCancel}>Cancel</button>
            <button data-testid="delete-confirmation-dialog-delete-btn" className={`btn ${primaryBtnClass}`} onClick={onPrimary}>{primaryBtnText}</button>
        </Action>
    </Modal>
}