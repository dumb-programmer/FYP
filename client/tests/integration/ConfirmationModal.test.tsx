import React, { useRef } from "react";
import { render, screen } from "@testing-library/react";
import { expect, describe, it, vi, beforeEach, afterEach } from "vitest";
import ConfirmationModal from "../../src/components/ConfirmationModal";

const mockedOnDelete = vi.fn();
vi.mock("react", async () => ({ ...await vi.importActual("react"), useRef: vi.fn() }));

const ConfirmationModalWrapper = () => {
    const ref = useRef<HTMLDialogElement | null>(null);

    return <ConfirmationModal dialogRef={ref} title="Are you sure?" description="This is a non-recoverable action." onPrimary={mockedOnDelete} onCancel={() => ref.current?.close()} />;
};

const renderComponent = () => render(<ConfirmationModalWrapper />);

describe("ConfirmationModal", () => {
    beforeEach(() => {
        HTMLDialogElement.prototype.showModal = vi.fn(function () {
            this.style.display = 'block';
            this.setAttribute('open', '');
        });

        HTMLDialogElement.prototype.close = vi.fn(function () {
            this.style.display = 'none';
            this.removeAttribute('open');
        });
    });

    afterEach(() => {
        vi.clearAllMocks()
    })

    it("renders correctly", () => {
        renderComponent();

        const modal = screen.getByTestId("modal");

        expect(modal).toMatchSnapshot();
    });
})