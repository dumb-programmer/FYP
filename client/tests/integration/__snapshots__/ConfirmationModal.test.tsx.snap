// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConfirmationModal > renders correctly 1`] = `
<dialog
  class="modal"
  data-testid="modal"
  id=":r0:"
>
  <div
    class="modal-box"
  >
    <h1
      class="text-xl font-bold"
    >
      Are you sure?
    </h1>
    <p
      class="mt-4"
    >
      This is a non-recoverable action.
    </p>
    <div
      class="modal-action"
    >
      <button
        class="btn"
        data-testid="delete-confirmation-dialog-cancel-btn"
      >
        Cancel
      </button>
      <button
        class="btn btn-error"
        data-testid="delete-confirmation-dialog-delete-btn"
      >
        Delete
      </button>
    </div>
  </div>
</dialog>
`;
